kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/ConfigurableFilterContributor.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter;
+
+import java.util.Objects;
+
+import io.kroxylicious.proxy.config.BaseConfig;
+import io.kroxylicious.proxy.service.ConfigurationDefinition;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A Convenience FilterContributor for Filters that require configuration.
+ * @param <T> The Configuration Type
+ */
+public abstract class ConfigurableFilterContributor<T extends BaseConfig> implements FilterContributor {
+
+    private final String typeName;
+    private final ConfigurationDefinition configDefiniton;
+    @NonNull
+    private final Class<T> configClazz;
+
+    protected ConfigurableFilterContributor(@NonNull String typeName, @NonNull Class<T> configClazz, boolean configRequired) {
+        this.configClazz = configClazz;
+        Objects.requireNonNull(typeName, "typeName was null");
+        Objects.requireNonNull(configClazz, "configClazz was null");
+        this.typeName = typeName;
+        this.configDefiniton = new ConfigurationDefinition(configClazz, configRequired);
+    }
+
+    @NonNull
+    @Override
+    public String getTypeName() {
+        return typeName;
+    }
+
+    @NonNull
+    @Override
+    public ConfigurationDefinition getConfigDefinition() {
+        return configDefiniton;
+    }
+
+    @NonNull
+    @Override
+    public Filter getInstance(FilterConstructContext context) {
+        try {
+            T config = configClazz.cast(context.getConfig());
```

Comment:
If this cast is the _raison d'etre_ for this class then I've got to question whether it's really worth it. I understand you've got the best of intentions but:

- You now have an interface and an abstract class, which means the filter author has to decide which to use.
- The possibility that the don't extend the class means you've not really removed the problem. 
- Every extra declaration is code which we're on the hook for in terms of API compatibility. In this case we now have a constructor which we need to worry about the compatibility of.

Did you consider making `FilterConstructContext` generic in the config type? That would allow us to do the type check, rather than putting on the filter author side.

Reply:
I'll have a go at moving the type-checking out of the Filter into the framework.